args/audio/systems/audiosystem.inl

```diff
@@ -0,0 +1,279 @@
+#pragma once
+
+namespace args::audio
+{
+	inline void AudioSystem::setup()
+	{
+		m_lock = new async::readonly_rw_spinlock();
+		async::readwrite_guard guard(*m_lock);
+
+		data::alDevice = alcOpenDevice(NULL);
+		if (!data::alDevice)
+		{
+			// Failed to create alcDevice
+			log::error("OpenAl device failed to open");
+#if defined(AUDIO_EXIT_ON_FAIL)
+			raiseEvent<events::exit>();
+#endif
+			return;
+		}
+		// Succesfully created alcDevice
+		log::info("Succesfully created openAl device");
+
+		data::alContext = alcCreateContext(data::alDevice, NULL);
+		if (!alcMakeContextCurrent(data::alContext))
+		{
+			// Failed to create alcContext
+			log::error("OpenAl context failed to create");
+			openal_error();
+#if defined(AUDIO_EXIT_ON_FAIL)
+			raiseEvent<events::exit>();
+#endif
+			return;
+		}
+		// Succesfully created alcContext
+		log::info("Succesfully created openAl context");
+
+		alListener3f(AL_POSITION, 0, 0, 0);
+		alListener3f(AL_VELOCITY, 0, 0, 0);
+		ALfloat ori[] = { 0, 0, 1.0f, 0, 1.0f, 0 };
+		alListenerfv(AL_ORIENTATION, ori);
+		m_listenerPosition = position(0,0,0);
+		log::info("initialized listener!");
+
+		alDopplerFactor(1.0f);
+		alDistanceModel(AL_EXPONENT_DISTANCE);
+
+		queryInformation();
+
+		//m_sourcePositions = std::unordered_map<ecs::component_handle<audio_source>, position>();
+
+		//ARGS function binding
+
+		sourceQuery = createQuery<audio_source>();
+
+		createProcess<&AudioSystem::update>("Audio");
+		bindToEvent<events::component_creation<audio_source>, &AudioSystem::onAudioSourceComponentCreate>();
+		bindToEvent<events::component_destruction<audio_source>, &AudioSystem::onAudioSourceComponentDestroy>();
+		bindToEvent<events::component_creation<audio_listener>, &AudioSystem::onAudioListenerComponentCreate>();
+		bindToEvent<events::component_destruction<audio_listener>, &AudioSystem::onAudioListenerComponentDestroy>();
+
+		// Release context on this thread
+		alcMakeContextCurrent(nullptr);
+	}
+
+	inline void AudioSystem::queryInformation()
+	{
+		const ALchar* vendor = alGetString(AL_VENDOR);
+		const ALchar* version = alGetString(AL_VERSION);
+		const ALchar* renderer = alGetString(AL_RENDERER);
+		const ALchar* openALExtensions = alGetString(AL_EXTENSIONS);
+		const ALchar* ALCExtensions = alcGetString(data::alDevice, ALC_EXTENSIONS);
+		ALCint srate;
+		alcGetIntegerv(data::alDevice, ALC_FREQUENCY, 1, &srate);
+		log::info("OpenAL info:\n\n\t\t\tOpenAL information\n\tVendor: {}\n\tVersion: {}\n\tRenderer: {}\n\tOpenAl Extensions: {}\n\tALC Extensions: {}\n\tDevice samplerate: {}\n",
+			vendor, version, renderer, openALExtensions, ALCExtensions, srate);
+	}
+
+	inline void AudioSystem::update(time::span deltatime)
+	{
+		alcMakeContextCurrent(data::alContext);
```

Comment:
what if making the context current fails because another thread is utilizing the context?